cEP-0021.md

```diff
@@ -0,0 +1,215 @@
+| Status   | Proposed                                          |
+| Type     | Process                                           |
+
+## Abstract
+
+This cEP describes the detailed implementation of a profiling interface as a
+part of the
+[GSoC'18 project](https://summerofcode.withgoogle.com/projects/#6109762077327360).
+
+## Introduction
+
+Profiling a Python program is a form of dynamic analysis that measures the
+execution time of the program and the functions it is composed of. Its goal is
+to produce data about where program is spending time and what area might
+be worth optimizing. Thus we can provide a *Profiling interface* from within
+coala and make it available to bear writers so that they can create better
+performant bears.
+
+This cEP proposes to introduce the profiling interface so that bear writers
+will have the ability to Profile the Bear's code to optimize its
+performance.
+
+## Proposed Change
+
+Bear writers can invoke the profiler with the `--profile-bears` argument.
+
+Bear writer will have the ability to directly dump the raw
+profile output (binary output) either on *current working directory* or to a
+*specified directory name*, which can be further used for examination
+of profiler stats with the help of different modules like
+`pstats` or `snakeviz`.
+
+- `--profile-bears` or `profile_bears` (using `.coafile`) is the main
+  argument to enable profiling. It accepts an additional parameter
+  `Dirname` through which bear writers can specify where to store the dump.
+
+The profiler will start by profiling the `run()` method of bears
+because it consumes most of the bears time. So, this is part where bear writer
+will spend time, as rest of the part like loading the files, collecting the
+settings, etc. are done by coala itself.
+
+## Implementation Details
+
+When user invokes the profiler with either *CLI* or by `coafile`
+then the `execute` method of Bear base class will call the
+`run_bear_from_section` method.
+
+```python
+def run_bear_from_section(self, args, kwargs):
+  ...
+  profile_bears = kwargs.get('profile_bears')
+  if profile_bears is None:
+      self.run(*args, **kwargs)
+  else:
+      self.profile(self.run, *args, **kwargs)
+  ...
+```
+
+`run_bear_from_section` method checks whether bear writer invoke the profiler
+interface or not i.e., keyword arguments passed to the Bear contain the
+`profile_bears` key or not.
+
+- If *not* then `run_bear_from_section()` method will directly call
+  the `run()` method of Bear so that coala can continue its normal
+  execution.
+- Else `run_bear_from_section()` method will call `profile()` method which
+  will check the whether the value of `profile_bears` key value is
+  `True` or `Dirname`.
+
+  - If `True` then `profile()` will create a profiler file for every
+    bear of every section with a name `{section_name}_{bear_name}.prof` and
+    dump it in a *current working directory.*
+
+    ```python
+    path = os.getcwd()
+    if str(profile_bears)=='True':
```

Comment:
this wasnt run through pycodestyle :P

Reply:
Sorry for late response, making changes right not :+1: